vms/platformvm/txs/mempool/mempool.go

```diff
@@ -5,42 +5,411 @@ package mempool
+	maxHeap            heap.Map[ids.ID, Tx]
+	minHeap            heap.Map[ids.ID, Tx]
+	consumedUTXOs      *setmap.SetMap[ids.ID, ids.ID]
+	droppedTxIDs       *lru.SizedCache[ids.ID, error]
+	currentGas         gas.Gas
+	numTxsMetric       prometheus.Gauge
+	gasAvailableMetric prometheus.Gauge
 }
 
-func New(namespace string, registerer prometheus.Registerer) (*Mempool, error) {
-	metrics, err := txmempool.NewMetrics(namespace, registerer)
-	if err != nil {
-		return nil, err
+func New(
+	namespace string,
+	weights gas.Dimensions,
+	gasCapacity gas.Gas,
+	avaxAssetID ids.ID,
+	registerer prometheus.Registerer,
+) (*Mempool, error) {
+	numTxsMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "count",
+		Help:      "number of transactions in the mempool",
+	})
+
+	gasAvailableMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: namespace,
+		Name:      "gas_available",
+		Help:      "amount of remaining gas available",
+	})
+
+	if err := errors.Join(
+		registerer.Register(numTxsMetric),
+		registerer.Register(gasAvailableMetric),
+	); err != nil {
+		return nil, fmt.Errorf("failed to register metrics: %w", err)
 	}
-	pool := txmempool.New[*txs.Tx](
-		metrics,
-	)
-	return &Mempool{Mempool: pool}, nil
+
+	return &Mempool{
+		weights:     weights,
+		gasCapacity: gasCapacity,
+		avaxAssetID: avaxAssetID,
+		maxHeap: heap.NewMap[ids.ID, Tx](func(a, b Tx) bool {
+			return a.GasPrice > b.GasPrice
+		}),
+		minHeap: heap.NewMap[ids.ID, Tx](func(a, b Tx) bool {
+			return a.GasPrice < b.GasPrice
+		}),
+		consumedUTXOs: setmap.New[ids.ID, ids.ID](),
+		droppedTxIDs: lru.NewSizedCache[ids.ID, error](64, func(ids.ID, error) int {
+			return 1
+		}),
+		numTxsMetric:       numTxsMetric,
+		gasAvailableMetric: gasAvailableMetric,
+	}, nil
 }
 
 func (m *Mempool) Add(tx *txs.Tx) error {
-	switch tx.Unsigned.(type) {
-	case *txs.AdvanceTimeTx:
-		return ErrCantIssueAdvanceTimeTx
-	case *txs.RewardValidatorTx:
-		return ErrCantIssueRewardValidatorTx
-	default:
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	complexity, gasUsed, err := m.meter(tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to meter tx: %w", err)
+	}
+
+	consumedAVAX, producedAVAX, err := getUTXOs(m.avaxAssetID, tx.Unsigned)
+	if err != nil {
+		return fmt.Errorf("failed to get utxos: %w", err)
+	}
+
+	feesPaid := consumedAVAX - producedAVAX
+	gasPrice := float64(0)
+	if gasUsed == 0 {
+		gasPrice = 0
+	} else {
+		gasPrice = float64(feesPaid) / float64(gasUsed)
+	}
+
+	// Evict a lower paying tx if we do not have enough remaining gas capacity
+	if next := m.currentGas + gasUsed; next > m.gasCapacity {
+		if err := m.tryEvictTx(next, gasPrice); err != nil {
+			return err
+		}
+	}
+
+	heapTx := Tx{
+		Tx:         tx,
+		Complexity: complexity,
+		gasUsed:    gasUsed,
+		GasPrice:   gasPrice,
+	}
+
+	if m.consumedUTXOs.HasOverlap(tx.InputIDs()) {
+		return fmt.Errorf("failed to add tx: %w", mempool.ErrConflictsWithOtherTx)
+	}
```

Comment:
Should this be done before evicting the lower price tx?

Reply:
Yeah

Reply:
Yeah... I think we also need to check if we're conflicting with the transaction we're evicting, in which case it's fine to add